lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+require 'kubernetes-deploy/kubernetes_resource'
+require 'kubernetes-deploy/kubeclient_builder'
+require 'erb'
+require 'json'
+require 'open3'
+
+module KubernetesDeploy
+  class DiscoverableResource < KubernetesResource
+    extend KubernetesDeploy::KubeclientBuilder
+
+    TRUE_VALUES = [true, 1, '1', 't', 'T', 'true', 'TRUE'].to_set
+    DEPLOY_METADATA_ANNOTATION = 'kubernetes-deploy.shopify.io/metadata'
+
+    def self.inherited(child_class)
+      DiscoverableResource.child_classes.add(child_class)
+    end
+
+    def self.discover(context:, logger:, server_version:)
+      logger.info("Discovering custom resources:")
+      with_retries { discover_groups(context) }
+      if server_version >= Gem::Version.new('1.7.0')
+        kube_client = v1beta1_crd_kubeclient(context)
+        with_retries { discover_crd(logger: logger, client: kube_client) }
+      end
+    end
+
+    def self.discover_groups(context)
+      kinds = discover_kinds(context)
+      kinds.each_pair do |key, val|
+        klass = get_static_class(kind: key)
+        next unless klass
+        klass.const_set(:GROUP, val[:group]) unless klass.constants.include?(:GROUP)
+        klass.const_set(:VERSION, val[:version]) unless klass.constants.include?(:VERSION)
+      end
+    end
+
+    def self.discover_kinds(context)
+      kinds = {}
+
+      # At the top level there is the core group (everything below /api/v1),
```

Comment:
Unfortunately for historic reasons not all resources are available under `/apis/$NAME/$VERSION)`, so we have to look in two places.